src/app/nodetool/qgsnodetool2.cpp

```diff
@@ -0,0 +1,1393 @@
+class MatchCollectingFilter : public QgsPointLocator::MatchFilter
+{
+  public:
+    QList<QgsPointLocator::Match> matches;
+    QgsNodeTool2* nodetool;
+
+    MatchCollectingFilter( QgsNodeTool2* nodetool )
+        : nodetool( nodetool ) {}
+
+    virtual bool acceptMatch( const QgsPointLocator::Match& match ) override
+    {
+      if ( match.distance() > 0 )
+        return false;
+      matches.append( match );
+
+      // there may be multiple points at the same location, but we get only one
+      // result... the locator API needs a new method verticesInRect()
+      QgsGeometry matchGeom = nodetool->cachedGeometry( match.layer(), match.featureId() );
+      QgsVertexId vid;
+      QgsPointV2 pt;
+      while ( matchGeom.geometry()->nextVertex( vid, pt ) )
+      {
+        int vindex = matchGeom.vertexNrFromVertexId( vid );
+        if ( pt.x() == match.point().x() && pt.y() == match.point().y() && vindex != match.vertexIndex() )
+        {
+          QgsPointLocator::Match extra_match( match.type(), match.layer(), match.featureId(),
+                                              0, match.point(), vindex );
+          matches.append( extra_match );
+        }
+      }
+      return true;
+    }
+};
+
+//
+//
+//
+
+
+QgsNodeTool2::QgsNodeTool2( QgsMapCanvas *canvas, QgsAdvancedDigitizingDockWidget *cadDock )
+    : QgsMapToolAdvancedDigitizing( canvas, cadDock )
+{
+  mSnapMarker = new QgsVertexMarker( canvas );
+  mSnapMarker->setIconType( QgsVertexMarker::ICON_CROSS );
+  mSnapMarker->setColor( Qt::magenta );
+  mSnapMarker->setPenWidth( 3 );
+  mSnapMarker->setVisible( false );
+
+  mEdgeCenterMarker = new QgsVertexMarker( canvas );
+  mEdgeCenterMarker->setIconType( QgsVertexMarker::ICON_CROSS );
+  mEdgeCenterMarker->setColor( Qt::red );
+  mEdgeCenterMarker->setPenWidth( 3 );
+  mEdgeCenterMarker->setVisible( false );
+
+  mDragPointMarker = new QgsVertexMarker( canvas );
+  mDragPointMarker->setIconType( QgsVertexMarker::ICON_X );
+  mDragPointMarker->setColor( Qt::red );
+  mDragPointMarker->setPenWidth( 3 );
+  mDragPointMarker->setVisible( false );
+
+  mFeatureBand = createRubberBand( QgsWkbTypes::LineGeometry );
+  mFeatureBand->setVisible( false );
+
+  QColor color = digitizingStrokeColor();
+  mVertexBand = new QgsRubberBand( canvas );
+  mVertexBand->setIcon( QgsRubberBand::ICON_CIRCLE );
+  mVertexBand->setColor( color );
+  mVertexBand->setIconSize( 15 );
+  mVertexBand->setVisible( false );
+
+  QColor color2( color );
+  color2.setAlpha( color2.alpha() / 3 );
+  mEdgeBand = new QgsRubberBand( canvas );
+  mEdgeBand->setColor( color2 );
+  mEdgeBand->setWidth( 10 );
+  mEdgeBand->setVisible( false );
+
+  mEndpointMarker = new QgsVertexMarker( canvas );
+  mEndpointMarker->setIconType( QgsVertexMarker::ICON_CROSS );
+  mEndpointMarker->setColor( Qt::red );
+  mEndpointMarker->setPenWidth( 3 );
+  mEndpointMarker->setVisible( false );
+}
+
+QgsNodeTool2::~QgsNodeTool2()
+{
+  delete mSnapMarker;
```

Comment:
I see your point for a tool like this, but I think it's just a good habit for us to get into. Plus, we still want to be sure we don't leak even with objects which are bound to the lifetime of QGIS as it avoids valgrind noise.